samples/batch/batch-ingestion-client/RealtimeTranscription/Helpers/RealtimeTranscriptionHelper.cs

```diff
@@ -0,0 +1,119 @@
+
+            var jsonResults = new List<JsonResult>();
+
+            var chunkId = Encoding.ASCII.GetString(audio.Take(4).ToArray());
+
+            // Verify that first 4 bytes are RIFF in ascii:
+            // (see https://docs.fileformat.com/audio/wav/ for details)
+            if (chunkId != "RIFF")
+            {
+                throw new InvalidOperationException($"Expected file header containing RIFF, received {chunkId} instead.");
+            }
+
+            // Get bytes for sampleRate, bitsPerSample, and channels:
+            var sampleRate = BitConverter.ToUInt32(audio.Skip(24).Take(4).ToArray());
+            var bitsPerSample = audio[34];
+            var channels = audio[22];
+
+            var audioStreamFormat = AudioStreamFormat.GetWaveFormatPCM(sampleRate, bitsPerSample, channels);
+
+            var stopRecognition = new TaskCompletionSource<int>();
+            using var pushStream = AudioInputStream.CreatePushStream(audioStreamFormat);
+            using (var audioInput = AudioConfig.FromStreamInput(pushStream))
+            {
+                using var recognizer = new SpeechRecognizer(
+                    speechConfig,
+                    audioInput);
+
+                using var connection = Connection.FromRecognizer(recognizer);
+
+                pushStream.Write(audio);
+                pushStream.Close();
+
+                logger.LogInformation("Starting speech recognition.");
+                recognizer.Recognized += (s, e) =>
+                {
+                    if (e.Result.Reason == ResultReason.RecognizedSpeech)
+                    {
+                        var stringResult = e.Result.Properties.GetProperty(PropertyId.SpeechServiceResponse_JsonResult);
+                        var deserializedResult = JsonConvert.DeserializeObject<JsonResult>(stringResult);
+                        jsonResults.Add(deserializedResult);
+                    }
+                };
+
+                recognizer.SessionStarted += (s, e) =>
+                {
+                    logger.LogInformation($"Starting session. Session id: {e.SessionId}.");
+                };
+
+                recognizer.SessionStopped += (s, e) =>
+                {
+                    logger.LogInformation("Session stopped");
+                    stopRecognition.TrySetResult(0);
+                };
+
+                recognizer.Canceled += (s, e) =>
+                {
+                    var cancellation = CancellationDetails.FromResult(e.Result);
+                    logger.LogError($"Recognition canceled: Reason={cancellation.Reason}");
+
+                    if (cancellation.Reason == CancellationReason.Error)
+                    {
+                        logger.LogError($"ErrorCode={cancellation.ErrorCode}");
+                        logger.LogError($"ErrorDetails={cancellation.ErrorDetails}");
+                    }
+
+                    stopRecognition.TrySetResult(0);
```

Comment:
What happens if the file has not been completed transcribed when you get canceled event? do you want to start a new session to continue left audios?

Reply:
throwing an exception and retrying based on the error now!